KAS-3992: Add shortlist for publication-flows

sergiofenoll opened the pull request:
JIRA: https://kanselarij.atlassian.net/browse/KAS-3992

Adds a new route to the publications page where users can find a list of pieces that need to have their publications started.
![image](https://user-images.githubusercontent.com/22411874/231115595-ebae9e5f-dffe-46e0-8218-db5beb41af1e.png)

Related PRs:
- [x] https://github.com/kanselarij-vlaanderen/app-kaleidos/pull/390

Related service:
- [x] https://github.com/kanselarij-vlaanderen/shortlist-service

sergiofenoll committed:
feat: add shortlist for publication-flows

sergiofenoll committed:
chore: remove unused helper from shortlit controller

sergiofenoll committed:
style: apply prettier to shortlist template

sergiofenoll committed:
feat: fetch shortlit piece data from resources

Instead of getting all the data from the service, only get the UUIDs and
then fetch the actual data from mu-cl-resources. This prevents issues
with our records having incomplete data (because the service only
returns the pertinent data about e.g. the latest agendaitem).

sergiofenoll committed:
chore: remove console.timeEnd statement

sergiofenoll committed:
fix: handle empty responses from the shortlist service

sergiofenoll committed:
fix: await async call

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3992-publications-shortlist

sergiofenoll committed:
Merge branch 'development' into feature/KAS-3992-publications-shortlist

ValenberghsSven committed:
Merge branch 'development' of github.com:kanselarij-vlaanderen/frontend-kaleidos into feature/KAS-3992-publications-shortlist

ValenberghsSven commented on app/controllers/publications/overview/shortlist.js:
```diff
@@ -0,0 +1,127 @@
+  caseForPublication;
+  mandateesForPublication;
+
+  getAgendaitem = async (piece) => {
+    const agendaitems = await piece.agendaitems;
+    let agendaitem;
+    for (let maybeAgendaitem of agendaitems) {
+      const agenda = await maybeAgendaitem.agenda;
+      const nextVersion = await agenda.nextVersion;
+      if (!nextVersion) {
+        agendaitem = maybeAgendaitem;
+        break;
+      }
+    }
+    return agendaitem;
+  }
```
Prefer a query for this one.
We are loading a lot of data that we eventually don't need.

```
await this.store.queryOne('agendaitem', {
      'filter[pieces][:id:]': this.args.piece.id,
      'filter[:has-no:next-version]': 't',
      sort: '-created',
```
This will yield the latest agendaitem linked to that piece.

sergiofenoll replied:
Because all the data is included, the only calls we actually do is getting `agendaitem.nextVersion` in the case that the nextVersion is null (and we could probably even work around this and have 0 calls instead).

Using a query means that we're doing this call over and over again, because `store.query(One)` isn't cached.
I'm okay with re-writing this but this will turn this route into one of many routes where we're making 10s or 100s of calls just to show a list.

ValenberghsSven commented on app/controllers/publications/overview/shortlist.js:
```diff
@@ -0,0 +1,127 @@
+          CONSTANTS.REGULATION_TYPES.BVR
+        );
+        break;
+      case CONSTANTS.DOCUMENT_TYPES.MB:
+        regulationType = await this.store.findRecordByUri(
+          'regulation-type',
+          CONSTANTS.REGULATION_TYPES.MB
+        );
+        break;
```
MB doc type should not exist here? 
Shortlist service only returns Decreet and BVR
If any other doc type is returned, does that mean we should abort creating the publication instead of setting it undefined?

sergiofenoll replied:
I copy-pasted the logic for creating publications, this case can just be removed. Will do.

ValenberghsSven commented on app/routes/publications/overview/shortlist.js:
```diff
@@ -0,0 +1,33 @@
+    const result = await response.json();
+
+    if (result.data.length) {
+      return this.store.query('piece', {
+        include: [
+          'agendaitems.agenda.next-version',
+          'agendaitems.mandatees.person',
+          'agendaitems.treatment.decision-activity',
+          'document-container.type',
+        ].join(','),
+        sort: '-created',
+        'page[size]': result.data.length,
+        'filter[:id:]': result.data.map((record) => record.id).join(','),
+      });
+    }
```
Potentially loads too much data (all mandatees for all agendaitem versions linked to 1 piece)
But we only need the latest agendaitem.
Perhaps this one should be simpler with just including `document-container.type`.
We could then just rely on the getter (waiting in the template) or load some data in the afterModel instead

sergiofenoll replied:
Same remark as above. We can remove the includes (but with new mu-cl-resources this call is relatively fast), but that also means we're going to be making a lot of calls when rendering. Best to check if this call is fast enough first (it is, on my machine).

ValenberghsSven commented on app/controllers/publications/overview/shortlist.js:
```diff
@@ -0,0 +1,127 @@
+          regulationType: regulationType,
+        }
+      );
+    this.pieceForPublication.publicationFlow = publicationFlow;
+    await this.pieceForPublication.save();
```
Saving the piece here.
We do this for publications I see (this is also in older code).
This could lead to concurrency issues.
Perhaps user 1 is starting a pub-flow while user 2 is updating the access right to public.
Also, with the default save override we also change `modified` and `accessLevelLastModified` properties by doing this.
Could we just create `pub-flow` with a list (of 1) instead so we set the relation on the inverse?

sergiofenoll replied:
Many-to-one relations aren't serialized by default by Ember Data, so we'd have to adapt the serializer for that to work (which is not a problem). But I don't see how that resolves the issue?

If I save a publication-flow with one linked piece and then another user saves that same piece (which has stale data, i.e. does not have the relationship set to the publication-flow), we still have the same problem.

ValenberghsSven replied:
Both have potential risk.
Sometimes there are pieces that have lost the connection to their file (after a replace).
This could be one of the causes with concurrency.
Perhaps we should consider using `ModelWithModifier` as a concurrency check for editing pieces and act accordingly
Also, `modified` and `accessLevelLastModified` always being saved might need some change.
Out of scope here.

ValenberghsSven commented on app/controllers/publications/overview/shortlist.js:
```diff
@@ -0,0 +1,127 @@
+  @service store;
+
+  @tracked isOpenNewPublicationModal;
+  pieceForPublication;
+  agendaitemForPublication;
+  decisionActivityForPublication;
+  caseForPublication;
+  mandateesForPublication;
```
should these be tracked? we are updating these constantly.
Or not needed if we don't use them in the template?

sergiofenoll replied:
Only used in JS so tracking not necessary as far as I understand.

sergiofenoll committed:
chore: remove unused branch in switch case

The shortlist only contains BVRs and Decreten, so no need for the branch
for MBs.

ValenberghsSven committed:
KAS-3992 add dutch name to route

ValenberghsSven committed:
KAS-3992 fix shortlist page throwing error when service is down

ValenberghsSven reviewed (approved):
made a minor change (route was not showing "op-te-starten" in address)
and error (page not opening) if service was not running.
